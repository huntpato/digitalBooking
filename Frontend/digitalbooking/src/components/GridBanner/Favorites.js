import React, { useEffect, useState, useContext } from "react";
import { UserContext } from "../UserContext/UserContext";
import styles from "./GridBanner.module.css";

export const Favorites = (props) => {
  const { notFav, trueFav } = styles;
  const [favData, setFavData] = useState([]);
  const [fav, setFav] = useState(false);
  const [statId, setStatId] = useState();
  const { productId } = props;
  const { user } = useContext(UserContext);
  const userId = user && JSON.parse(user).id;
  const bearer = user && `${JSON.parse(localStorage.getItem("bearer")).jwt}`;
  const options = {
    headers: {
      Authorization: bearer,
      "Content-Type": "application/json",
      "X-Requested-With": "XMLHttpRequest",
    },
  };
  const url =
    user &&
    `https://cors-everywhere-me.herokuapp.com/g4bookingapp-env-2.eba-4mbmxg4f.us-east-1.elasticbeanstalk.com/users/${userId}`;
  const url2 = `https://cors-everywhere-me.herokuapp.com/g4bookingapp-env-2.eba-4mbmxg4f.us-east-1.elasticbeanstalk.com/stats/save`;
  const url3 = `https://cors-everywhere-me.herokuapp.com/g4bookingapp-env-2.eba-4mbmxg4f.us-east-1.elasticbeanstalk.com//stats/update`;
  let form = {
    user: { id: userId },
    product: { id: productId },
    like: !fav,
  };
  let form2= {
    id: statId,
    user: { id: userId },
    product: { id: productId },
    like: !fav,
  };

  useEffect(() => {
    if (user) {
      fetch(url, options)
        .then((response) => response.json())
        .then((data) => {
          setFavData(data.stats);
        })
        .catch((error) => {
          console.log(error);
        });
    }
  }, [user]);

  useEffect(() => {
    favData.map((stat) => {
      if(stat.productId == productId){
        setStatId(stat.id);
      }
      if (stat.productId == productId && stat.like) {
        setFav(true);
      }
    });
  }, [favData]);

  const handleFav = (e) => {
    e.preventDefault();
    if (!fav && statId == undefined) {
      fetch(url2, {
        method: "POST",
        headers: {
          Authorization: bearer,
          "Content-Type": "application/json",
          "X-Requested-With": "XMLHttpRequest",
        },
        body: JSON.stringify(form),
      })
        .then((response) => response.json())
        .then((data) => setStatId(data.id))
        .catch((error) => {
          console.log(error);
        });
    } else if (!fav) {
      fetch(url3, {
        method: "PUT",
        headers: {
          Authorization: bearer,
          "Content-Type": "application/json",
          "X-Requested-With": "XMLHttpRequest",
        },
        body: JSON.stringify(form2),
      })
        .then((response) => response.json())
        .then((data) => console.log(data))
        .catch((error) => {
          console.log(error);
        });
    } else if (fav) {
      fetch(url3, {
        method: "PUT",
        headers: {
          Authorization: bearer,
          "Content-Type": "application/json",
          "X-Requested-With": "XMLHttpRequest",
        },
        body: JSON.stringify(form2),
      })
        .then((response) => response.json())
        .then((data) => console.log(data))
        .catch((error) => {
          console.log(error);
        });
    }
    setFav(!fav);
  };

  return (
    <div>
      {fav ? (
        <div className={trueFav}>
          <i className="fas fa-heart" onClick={handleFav}></i>
        </div>
      ) : (
        <div className={notFav}>
          <i className="fal fa-heart" onClick={handleFav}></i>
        </div>
      )}
    </div>
  );
};
