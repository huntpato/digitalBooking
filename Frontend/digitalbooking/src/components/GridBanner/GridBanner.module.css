.grid {
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  margin-bottom: 30px;
  width: 100%;
  /*background-color: darkseagreen;*/
}

.gridContainer {
  display: flex;
  flex-direction: column;
  width: 90vw;
  /*background-color: deepskyblue;*/
}

/*social icons section*/
.social {
  align-self: flex-start;
  display: flex;
  flex-direction: row;
  padding: 30px 0px 20px 0px;
}

.social i {
  color: var(--color-typo);
  font-family: "Font Awesome 5 Free";
  margin: 0;
  padding-right: 10px;
}

.shareIcon{
  cursor: pointer;
  transition: 0.9s;
}

.shareIcon:active{
  -webkit-animation: scale-up-center 0.4s cubic-bezier(0.39, 0.575, 0.565, 1)
  both;
animation: scale-up-center 0.4s cubic-bezier(0.39, 0.575, 0.565, 1) both;
}

.notFav {
  cursor: pointer;
  transition: 0.9s;
}

.notFav i:active {
  -webkit-animation: scale-up-center 0.8s cubic-bezier(0.39, 0.575, 0.565, 1)
    both;
  animation: scale-up-center 0.6s cubic-bezier(0.39, 0.575, 0.565, 1) both;
}

.trueFav i {
  cursor: pointer;
  color: var(--seconday-color);
  transition: 0.9s;
}

.trueFav i:active {
  -webkit-animation: scale-up-center 0.8s cubic-bezier(0.39, 0.575, 0.565, 1)
    both;
  animation: scale-up-center 0.6s cubic-bezier(0.39, 0.575, 0.565, 1) both;
}

/*title see more*/
.gridContainer h5 {
  color: white;
  cursor: pointer;
  text-decoration: underline;
  position: relative;
  align-self: flex-end;
  margin-right: 2%;
  z-index: 4;
  bottom: 30px;
}

/*images section*/
.image {
  align-self: center;
  display: grid;
  /*grid-template-columns: repeat(4, 1fr) 0;
  grid-template-rows: repeat(2, 1fr) repeat(3, 0);
  grid-column-gap: 5px;
  row-gap: 1vmin;*/
  grid-template-columns: repeat(4, 1fr);
  grid-template-rows: repeat(2, 1fr);
  grid-column-gap: 5px;
  grid-row-gap: 5px;
  padding: 0px;
  width: 100%;
}

.div1 {
  grid-area: 1 / 1 / 3 / 3;
  height: 100%;
  width: 100%;
}

.div1 img {
  border-radius: 10px 0 0 10px;
  object-fit: cover;
  width: 100%;
  height: 100%;
  max-height: 100%;
}

.div2 {
  grid-area: 1 / 3 / 2 / 4;
  width: 100%;
  height: 100%;
  object-position: 100% 0;
}

.div2 img {
  object-fit: cover;
  width: 100%;
  height: 100%;
  max-height: 100%;
}

.div3 {
  grid-area: 1 / 4 / 2 / 5;
  width: 100%;
  height: 100%;
  object-position: 100% 0;
}

.div3 img {
  border-radius: 0 10px 0 0;
  object-fit: cover;
  width: 100%;
  height: 100%;
  max-height: 100%;
}

.div4 {
  grid-area: 2 / 3 / 3 / 4;
  width: 100%;
  height: 100%;
  object-position: 100% 0;
}

.div4 img {
  object-fit: cover;
  width: 100%;
  height: 100%;
  max-height: 100%;
}

.div5 {
  grid-area: 2 / 4 / 3 / 5;
  width: 100%;
  height: 100%;
  object-position: 100% 0;
}

.div5 img {
  border-radius: 0 0 10px 0;
  object-fit: cover;
  width: 100%;
  height: 100%;
  max-height: 100%;
}

.wrapper {
  background-color: #202020;
}

.share {
  position: absolute;
  z-index: 3;
}

@media (max-width: 768px) {
  .grid {
    align-items: center;
    display: flex;
    flex-direction: column;
    justify-content: center;
    margin-bottom: 0px;
    width: 100%;
    /*background-color: darkseagreen;*/
  }

  .social i {
    padding-right: 10px;
  }

  /*title see more*/
  .gridContainer h5 {
    bottom: 35px;
  }

  .image {
    display: grid;
    grid-template-columns: 1fr;
    grid-template-rows: 1fr;
    grid-column-gap: 0px;
    grid-row-gap: 0px;
    width: 100vw;
  }

  .div1 {
    grid-area: 1 / 1 / 2 / 2;
    height: 100%;
    width: 100%;
  }

  .div1 img {
    border-radius: 0px 0px 0px 0px;
    object-fit: cover;
    width: 100%;
    height: 100%;
    max-height: 100%;
  }

  .div2,
  .div3,
  .div4,
  .div5 {
    display: none;
  }

  .image {
    display: none;
  }

  .gridContainer h5 {
    display: none;
  }
}

@media (max-width: 480px) {
  .gridContainer {
    position: absolute;
  }
  /*social icons section*/
  .social {
    padding: 0px 0px 0px 0px;
    position: absolute;
    top: 20px;
    z-index: 6;
  }

  .social i {
    color: white;
    font-size: 14px;
    padding-right: 10px;
  }
}

@-webkit-keyframes scale-up-center {
  0% {
    -webkit-transform: scale(0.5);
    transform: scale(0.5);
  }
  100% {
    -webkit-transform: scale(1);
    transform: scale(1);
  }
}
@keyframes scale-up-center {
  0% {
    -webkit-transform: scale(0.5);
    transform: scale(0.5);
  }
  100% {
    -webkit-transform: scale(1);
    transform: scale(1);
  }
}

