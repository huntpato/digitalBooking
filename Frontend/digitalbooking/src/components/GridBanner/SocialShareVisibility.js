import { createContext } from "react";

export const SocialShareVisibilty  = createContext({
    
});