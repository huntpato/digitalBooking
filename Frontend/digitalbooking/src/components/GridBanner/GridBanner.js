import React, {useState, useMemo, useContext} from "react";
import { GalleryContext } from "../ProductService/GalleryContext";
import { SocialShare } from "../SocialShare/SocialShare";
import styles from "./GridBanner.module.css";
import { SocialShareVisibilty } from "./SocialShareVisibility";
import { Favorites } from "./Favorites";

export const GridBanner = (props) => {
  const { div1, shareIcon, div2, div3, div4, div5, grid, gridContainer, image, social, share } = styles;
  const {setGallery}= useContext(GalleryContext);
  const { ph1, ph2, ph3, ph4, ph5, productId} = props;
  const [visibility, setVisibility] = useState(false);
  const value = useMemo(() => ({ visibility, setVisibility }), [visibility, setVisibility ]);
 

  const handleClick= ()=>{
    setVisibility(true)
  }


  const handleGallery =()=>{
    setGallery(true)
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  }


  return (
    <div className={grid}>
      <div className={gridContainer}>
        <div className={social}>
          <div className={shareIcon}><i className="fas fa-share-alt" onClick={handleClick}></i></div>
          <Favorites productId={productId} />
          {visibility &&  <SocialShareVisibilty.Provider value={value}>
            <SocialShare className={share}/> </SocialShareVisibilty.Provider>}
        </div>
        <div className={image}>
          <div className={div1}><img src={ph1} alt="imagen"/></div>
          <div className={div2}><img src={ph2} alt="imagen"/></div>
          <div className={div3}><img src={ph3} alt="imagen"/></div>
          <div className={div4}><img src={ph4} alt="imagen"/></div>
          <div className={div5}><img src={ph5} alt="imagen"/></div>
        </div>
        <h5 onClick={handleGallery}>Ver m√°s</h5>
      </div>
    </div>
  );
};
