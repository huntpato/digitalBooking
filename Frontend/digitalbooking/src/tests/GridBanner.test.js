import React from 'react';
import { shallow } from 'enzyme';
import { GridBanner } from '../components/GridBanner/GridBanner'
import '@testing-library/jest-dom'
import { expect } from '@jest/globals';

describe('Pruebas en componente <GridBanner />', ()=> {
    const ph1="https://images.unsplash.com/photo-1440778303588-435521a205bc?ixid=MnwxMjA3fDB8MHxzZWFyY2h8NXx8aG9saWRheXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60";
    const ph2="https://images.unsplash.com/photo-1527179528411-4219e0714bcc?ixid=MnwxMjA3fDB8MHxzZWFyY2h8NHx8aG9saWRheXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60";
    const ph3="https://images.unsplash.com/photo-1473496169904-658ba7c44d8a?ixid=MnwxMjA3fDB8MHxzZWFyY2h8Nnx8aG9saWRheXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60";
    const ph4="https://images.unsplash.com/photo-1487349384428-12b47aca925e?ixid=MnwxMjA3fDB8MHxzZWFyY2h8OXx8aG9saWRheXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60";
    const ph5="https://images.unsplash.com/photo-1447501614729-24781f73c2f1?ixid=MnwxMjA3fDB8MHxzZWFyY2h8MTF8fGhvbGlkYXl8ZW58MHx8MHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60";
    const wrapper= shallow(<GridBanner ph1={ph1}  ph2={ph2}  ph3={ph3}  ph4={ph4}  ph5={ph5} />);

    test('Comprobar que <GridBanner /> renderiza correctamente.', () => {
        expect( wrapper ).toMatchSnapshot();
    })

    test('Verificar que <GridBanner /> muestra las imagenes enviadas por props', () => {
    
        expect (wrapper.find("img").at(0).prop("src")).toEqual(ph1);
        expect (wrapper.find("img").at(1).prop("src")).toEqual(ph2);
        expect (wrapper.find("img").at(2).prop("src")).toEqual(ph3);
        expect (wrapper.find("img").at(3).prop("src")).toEqual(ph4);
        expect (wrapper.find("img").at(4).prop("src")).toEqual(ph5);
    }) 

    test('Verificar que el componente carga la clase de CSS correctamente', () => {

        const gridBanner = wrapper.find('div').at(0);
        expect(gridBanner.hasClass('grid')).toBe(true);
    })

    test('Verificar que el componente carga el titulo h5 correctamente', () => {

        const gridBanner = wrapper.find('h5');
        expect(gridBanner.contains('Ver m√°s')).toBe(true);
    })
    

})
