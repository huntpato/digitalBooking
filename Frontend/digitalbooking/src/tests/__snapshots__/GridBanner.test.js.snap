// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente <GridBanner /> Comprobar que <GridBanner /> renderiza correctamente. 1`] = `
<div
  className="grid"
>
  <div
    className="gridContainer"
  >
    <div
      className="social"
    >
      <div
        className="shareIcon"
      >
        <i
          className="fas fa-share-alt"
          onClick={[Function]}
        />
      </div>
      <Favorites />
    </div>
    <div
      className="image"
    >
      <div
        className="div1"
      >
        <img
          alt="imagen"
          src="https://images.unsplash.com/photo-1440778303588-435521a205bc?ixid=MnwxMjA3fDB8MHxzZWFyY2h8NXx8aG9saWRheXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60"
        />
      </div>
      <div
        className="div2"
      >
        <img
          alt="imagen"
          src="https://images.unsplash.com/photo-1527179528411-4219e0714bcc?ixid=MnwxMjA3fDB8MHxzZWFyY2h8NHx8aG9saWRheXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60"
        />
      </div>
      <div
        className="div3"
      >
        <img
          alt="imagen"
          src="https://images.unsplash.com/photo-1473496169904-658ba7c44d8a?ixid=MnwxMjA3fDB8MHxzZWFyY2h8Nnx8aG9saWRheXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60"
        />
      </div>
      <div
        className="div4"
      >
        <img
          alt="imagen"
          src="https://images.unsplash.com/photo-1487349384428-12b47aca925e?ixid=MnwxMjA3fDB8MHxzZWFyY2h8OXx8aG9saWRheXxlbnwwfHwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60"
        />
      </div>
      <div
        className="div5"
      >
        <img
          alt="imagen"
          src="https://images.unsplash.com/photo-1447501614729-24781f73c2f1?ixid=MnwxMjA3fDB8MHxzZWFyY2h8MTF8fGhvbGlkYXl8ZW58MHx8MHx8&ixlib=rb-1.2.1&auto=format&fit=crop&w=500&q=60"
        />
      </div>
    </div>
    <h5
      onClick={[Function]}
    >
      Ver m√°s
    </h5>
  </div>
</div>
`;
